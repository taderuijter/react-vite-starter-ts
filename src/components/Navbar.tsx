import { Link } from "react-router-dom";

function Navbar() {
  return (
    <div
      data-testid="navbar"
      className="pt-4 pb-4 text-center bg-slate-200 mb-4">
      <div className="flex flex-col md:flex-row space-y-4 md:space-x-4 md:space-y-0 justify-center">
        <Link
          to="/"
          className="px-4 py-2 bg-white rounded-full hover:bg-blue-600 hover:text-white font-medium">
          Home
        </Link>
        <Link
          to="/about"
          className="px-4 py-2 bg-white rounded-full hover:bg-blue-600 hover:text-white font-medium">
          About
        </Link>
      </div>
    </div>
  );
}

export default Navbar;
