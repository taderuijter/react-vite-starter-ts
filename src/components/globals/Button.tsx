function Button() {
  return (
    <a
      href="https://github.com"
      target="_blank"
      rel="noopener noreferrer"
      className="inline-flex items-center justify-center px-6 py-2 text-sm font-semibold text-white transition-all duration-200 bg-primary border border-transparent rounded-md hover:bg-primary focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary"
      role="button"
      aria-label="GitHub repository">
      Check on GitHub
    </a>
  );
}

export default Button;
