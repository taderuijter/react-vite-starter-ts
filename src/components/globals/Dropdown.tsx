// Import React libraries
import { useState, useRef } from "react";
import {
  MagnifyingGlassIcon,
  ChevronDownIcon,
  ChevronUpIcon,
} from "@heroicons/react/20/solid";

// Custom hooks
import { useClickOutside } from "@/hooks/useClickOutside";

// Type checking for the dropdown
export type DropdownTypes = {
  search: boolean;
  options: { label: string; value: string }[];
  onChange: (value: string) => void;
};

// Component
export const Dropdown = ({ search, options, onChange }: DropdownTypes) => {
  const [label, setLabel] = useState(options[0].label);
  const [isOpen, setIsOpen] = useState(false);
  const [searchValue, setSearchValue] = useState("");
  const ref = useRef<HTMLDivElement>(null);

  const closeMenu = () => {
    setIsOpen(false);
  };

  useClickOutside(ref, closeMenu);

  const toggleMenu = () => {
    setIsOpen((prevOpen) => !prevOpen);
  };

  const dropdownHandler = (item: { label: string; value: string }) => {
    setLabel(item.label);
    onChange(item.value);
    setIsOpen(false);
    setSearchValue("");
  };

  const filteredDropdown = options.filter((item) =>
    item.label.toLowerCase().includes(searchValue.toLowerCase()),
  );

  return (
    <div className="relative inline-block text-left" ref={ref}>
      <div>
        <button
          type="button"
          className="flex rounded-md border px-4 py-2 space-x-2 border-gray-300 bg-white text-sm font-medium text-gray-700 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-primary transition-all duration-200"
          id="language-menu"
          aria-expanded="true"
          aria-haspopup="true"
          onClick={toggleMenu}>
          <div>{label}</div>
          {isOpen ? (
            <ChevronUpIcon className="w-[20px] h-[20px]" />
          ) : (
            <ChevronDownIcon className="w-[20px] h-[20px]" />
          )}
        </button>
      </div>
      {isOpen && (
        <div className="origin-top-right absolute right-0 mt-2 w-56 rounded-md shadow-lg bg-white ring-1 ring-black ring-opacity-5 focus:outline-none">
          <div className="py-1">
            {search && (
              <div className="flex items-center px-4 py-2">
                <label htmlFor="language-search" className="sr-only">
                  Search
                </label>
                <div className="relative">
                  <MagnifyingGlassIcon className="absolute top-0 left-0 w-[20px] h-[20px] ml-2 mt-2 text-gray-400" />
                  <input
                    type="text"
                    name="language-search"
                    id="language-search"
                    className="inline-flex items-center justify-center px-8 py-2 text-sm w-full rounded-md border border-gray-300 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-primary transition-all duration-200"
                    placeholder="Search"
                    value={searchValue}
                    onChange={(e) => setSearchValue(e.target.value)}
                  />
                </div>
              </div>
            )}

            {filteredDropdown.map((item) => (
              <button
                type="button"
                key={item.value}
                className={`${
                  label === item.label
                    ? "bg-gray-100 text-gray-900"
                    : "text-gray-700"
                } block px-4 py-2 text-sm w-full text-left focus:outline-none focus:ring-2 focus:ring-primary transition-all duration-200`}
                onClick={() => dropdownHandler(item)}>
                {item.label}
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default Dropdown;
