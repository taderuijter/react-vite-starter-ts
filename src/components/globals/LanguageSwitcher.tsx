import { useState, useRef } from "react";
import { useTranslation } from "react-i18next";
import {
  MagnifyingGlassIcon,
  ChevronDownIcon,
  ChevronUpIcon,
} from "@heroicons/react/20/solid";
import { useClickOutside } from "@/hooks/useClickOutside";

const LanguageSwitcher: React.FC = () => {
  const { t, i18n } = useTranslation();
  const [isOpen, setIsOpen] = useState(false);
  const [searchValue, setSearchValue] = useState("");
  const ref = useRef<HTMLDivElement>(null);

  const closeMenu = () => {
    setIsOpen(false);
  };

  useClickOutside(ref, closeMenu);

  const toggleMenu = () => {
    setIsOpen(!isOpen);
  };

  const setLanguage = (lang: string) => {
    i18n.changeLanguage(lang);
    setIsOpen(false);
    setSearchValue("");
  };

  const languageList = [
    { label: t("components.languageSwitcher.languages.en"), value: "en" },
    { label: t("components.languageSwitcher.languages.nl"), value: "nl" },
  ];

  const filteredList = languageList.filter((lang) =>
    lang.label.toLowerCase().includes(searchValue.toLowerCase()),
  );

  return (
    <div className="relative inline-block text-left" ref={ref}>
      <div>
        <button
          type="button"
          className="flex rounded-md border px-4 py-2 space-x-2 border-gray-300 bg-white text-sm font-medium text-gray-700 hover:bg-gray-50 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-offset-gray-100 focus:ring-indigo-500"
          id="language-menu"
          aria-expanded="true"
          aria-haspopup="true"
          onClick={toggleMenu}>
          <div>{t("components.languageSwitcher.currentLanguage")}</div>
          {isOpen ? (
            <ChevronUpIcon className="w-[20px] h-[20px]" />
          ) : (
            <ChevronDownIcon className="w-[20px] h-[20px]" />
          )}
        </button>
      </div>
      {isOpen && (
        <div className="origin-top-right absolute right-0 mt-2 w-56 rounded-md shadow-lg bg-white ring-1 ring-black ring-opacity-5 focus:outline-none">
          <div className="py-1">
            <div className="flex items-center px-4 py-2">
              <label htmlFor="language-search" className="sr-only">
                Search for a language
              </label>
              <div className="relative">
                <MagnifyingGlassIcon className="absolute top-0 left-0 w-[20px] h-[20px] ml-2 mt-2 text-gray-400" />
                <input
                  type="text"
                  name="language-search"
                  id="language-search"
                  className="focus:ring-black focus:border-black block w-full sm:text-sm border-gray-300 rounded-md pl-8"
                  placeholder="Search for a language"
                  value={searchValue}
                  onChange={(e) => setSearchValue(e.target.value)}
                />
              </div>
            </div>

            {filteredList.map((lang) => (
              <button
                type="button"
                key={lang.value}
                className={`${
                  i18n.language === lang.value
                    ? "bg-gray-100 text-gray-900"
                    : "text-gray-700"
                } block px-4 py-2 text-sm w-full text-left`}
                onClick={() => setLanguage(lang.value)}>
                {lang.label}
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default LanguageSwitcher;
