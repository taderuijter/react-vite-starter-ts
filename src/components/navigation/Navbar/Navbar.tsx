// Import React Libraries
import { useState } from "react";
import { Bars3Icon, XMarkIcon } from "@heroicons/react/20/solid";
import { Container } from "components/layout/Container/Container";

export interface NavbarProps {
  brand: React.ReactNode;
  navLinks?: React.ReactNode;
  style?: string;
}

export const Navbar = ({
  brand,
  navLinks,
  style = "bg-slate-100 py-4",
}: NavbarProps) => {
  const [expanded, setExpanded] = useState<boolean>(false);

  const toggleExpanded = (): void => {
    setExpanded(!expanded);
  };

  return (
    <header className={style}>
      <Container style="px-4 sm:px-6 lg:px-8">
        <div className="flex items-center justify-between">
          <div className="flex-shrink-0">
            <a href="/" title="Home" aria-label="Home" className="flex">
              {brand}
            </a>
          </div>

          <div className="flex lg:hidden">
            <button
              type="button"
              className="text-gray-900"
              aria-label="Toggle navigation"
              onClick={toggleExpanded}>
              {expanded ? (
                <XMarkIcon className="w-8 h-8" />
              ) : (
                <Bars3Icon className="w-8 h-8" />
              )}
            </button>
          </div>

          <nav className="hidden lg:items-center lg:ml-auto lg:space-x-8 lg:flex">
            {navLinks}
          </nav>
        </div>

        {expanded && (
          <nav>
            <div className="px-1 py-8">
              <div className="grid gap-y-7">{navLinks}</div>
            </div>
          </nav>
        )}
      </Container>
    </header>
  );
};
