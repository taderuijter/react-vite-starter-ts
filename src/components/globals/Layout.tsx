import Navbar from "@/components/globals/Navbar";

interface PropTypes {
  children: React.ReactNode;
}

function Layout({ children }: PropTypes) {
  return (
    <>
      <Navbar />
      {children}
    </>
  );
}

export default Layout;
