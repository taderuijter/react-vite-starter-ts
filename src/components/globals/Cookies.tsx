import { useContext } from "react";
import { Link } from "react-router-dom";
import { GlobalContext } from "@/store/GlobalContext";
import { Button } from "@/components/globals/Button";

export const Cookies = () => {
  const { state, dispatch } = useContext(GlobalContext);

  const handleClose = () => {
    dispatch({ type: "TOGGLE_COOKIES", payload: false });
  };

  return (
    <>
      {state.cookies && (
        <div className="pointer-events-none fixed inset-x-0 bottom-0 px-6 pb-6">
          <div className="pointer-events-auto ml-auto max-w-xl rounded-xl bg-white p-6 shadow-lg ring-1 ring-gray-900/10">
            <p className="text-sm leading-6 text-gray-900">
              Our website uses cookies to improve your browsing experience and
              to personalize the content and advertisements we show you. By
              using our website, you consent to our use of cookies. You can
              manage your cookie preferences at any time through your browser
              settings.
              <Link
                to="/privacy"
                className="underline underline-offset-4 ml-2 text-primary font-semibold hover:text-secondary">
                Cookie Policy
              </Link>
            </p>
            <div className="mt-4 flex items-center gap-x-2">
              <Button
                text="Accept all"
                type="button"
                styling="solid"
                onClick={handleClose}
              />
              <Button
                text="Reject all"
                type="button"
                styling="outline"
                onClick={handleClose}
              />
            </div>
          </div>
        </div>
      )}
    </>
  );
};
