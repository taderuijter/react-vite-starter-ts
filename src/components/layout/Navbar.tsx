// Import React Libraries
import { useState, useContext } from "react";
import { Link } from "react-router-dom";
import { Bars3Icon, XMarkIcon } from "@heroicons/react/20/solid";
import { useTranslation } from "react-i18next";
import { LanguageContext } from "@/store/LanguageContext";

// Import components
import { InternalLink, ExternalLink } from "@/components/globals/Links";
import { Dropdown } from "@/components/globals/Dropdown";

const Navbar = () => {
  const { t, i18n } = useTranslation();
  const [expanded, setExpanded] = useState<boolean>(false);
  const languages = useContext(LanguageContext);

  const toggleExpanded = (): void => {
    setExpanded(!expanded);
  };

  const changeLanguage = (value: string): void => {
    i18n.changeLanguage(value);
  };

  return (
    <header className="py-4 bg-slate-100">
      <div className="container px-4 mx-auto sm:px-6 lg:px-8">
        <div className="flex items-center justify-between">
          <div className="flex-shrink-0">
            <Link to="/" title="Home" aria-label="Home" className="flex">
              <span className="text-xl font-bold text-gray-900">
                React Boilerplate
              </span>
            </Link>
          </div>

          <div className="flex lg:hidden">
            <button
              type="button"
              className="text-gray-900"
              aria-label="Toggle navigation"
              onClick={toggleExpanded}>
              {expanded ? (
                <XMarkIcon className="w-8 h-8" />
              ) : (
                <Bars3Icon className="w-8 h-8" />
              )}
            </button>
          </div>

          <nav className="hidden lg:items-center lg:ml-16 lg:mr-auto lg:space-x-8 lg:flex">
            <InternalLink
              text={t("components.navbar.links.home")}
              href="/"
              styling="link"
            />
            <InternalLink
              text={t("components.navbar.links.about")}
              href="/about"
              styling="link"
            />
          </nav>

          <nav className="hidden lg:flex lg:items-center lg:justify-end lg:space-x-4">
            <Dropdown
              search={true}
              options={languages}
              onChange={(value) => {
                changeLanguage(value);
              }}
            />
            <ExternalLink
              text="Check on Github"
              href="https://www.github.com"
              styling="solid"
            />
          </nav>
        </div>

        {expanded && (
          <nav>
            <div className="px-1 py-8">
              <div className="grid gap-y-7">
                <InternalLink
                  text={t("components.navbar.links.home")}
                  href="/"
                  styling="link"
                />
                <InternalLink
                  text={t("components.navbar.links.about")}
                  href="/about"
                  styling="link"
                />

                <Dropdown
                  search={false}
                  options={languages}
                  onChange={(value) => {
                    changeLanguage(value);
                  }}
                />
                <ExternalLink
                  text="Check on Github"
                  href="https://www.github.com"
                  styling="solid"
                />
              </div>
            </div>
          </nav>
        )}
      </div>
    </header>
  );
};
export default Navbar;
