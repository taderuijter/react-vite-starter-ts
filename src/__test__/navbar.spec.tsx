import { describe, it, expect } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import Navbar from "@/components/Navbar";

describe("Navbar", () => {
  it("renders links to the Home and About pages", () => {
    render(
      <MemoryRouter>
        <Navbar />
      </MemoryRouter>,
    );
    const homeLink = screen.getByRole("link", { name: /home/i });
    const aboutLink = screen.getByRole("link", { name: /about/i });

    expect(homeLink.getAttribute("href")).toBe("/");
    expect(aboutLink.getAttribute("href")).toBe("/about");

    expect(screen.getByTestId("navbar")).toMatchSnapshot();
  });
});
