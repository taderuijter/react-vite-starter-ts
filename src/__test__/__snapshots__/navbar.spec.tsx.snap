// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar renders links to the Home and About pages 1`] = `
<div
  class="pt-4 pb-4 text-center bg-slate-200 mb-4"
  data-testid="navbar"
>
  <div
    class="flex flex-col md:flex-row space-y-4 md:space-x-4 md:space-y-0 justify-center"
  >
    <a
      class="px-4 py-2 bg-white rounded-full hover:bg-blue-600 hover:text-white font-medium"
      href="/"
    >
      Home
    </a>
    <a
      class="px-4 py-2 bg-white rounded-full hover:bg-blue-600 hover:text-white font-medium"
      href="/about"
    >
      About
    </a>
  </div>
</div>
`;
